feat: provider-agnostic Get Started docs

teamdandelion opened the pull request:
This commit converts the get started docs to show provider-agnostic API usage (via @llm.call), while still showing the specific code usage for each provider.

In the UI, I kept showing provider-specific code, both for consistency with the "Official SDK" comparison that's already included, but also because it's a very direct way of showing off Mirascope's wide range of supported providers.

Since the code is now nearly identical for each provider, we can unify to a single implementation (using openai gpt-4o-mini by default), and then substitute in provider-specific logic at docs build time. I've added logic for doing so (`generate_provider_examples.py`), and a pre-build hook that invokes it. The substitution logic is tested, and ensures that we get a build time error if the provider substitution fails (e.g. because an example didn't use openai/gpt-4o-mini as expected). Also, `mkdocs` is updated to watch the examples directory for file changes.

I haven't deleted any of the provider-specific implementations, since they're also used for `response_model.md`, and I wanted to get a minimal conversion out for review without leaving `response_model` in a half-switched state. However, if we move forward with this approach, we can delete most of the provider example files, leaving only cases like the sdk usage.

For cases where we don't have any provider-specific code to compare to, we could keep the per-provider tabbed structure, or we could drop it and just use openai or anthropic as consistent examples. I can see benefits to either approach, and I'm curious which you prefer.

Along the way, I made one API change, to expose `BaseCallResponse` in `mirascope.core`. This avoids bloating the example code with an extra import from `mirascope.core.base`, and seems very consistent with the other exports in `mirascope.core`. I think it'd be nice if we could get down to one import line in the standard case of using llm call, though I'm not sure if there's a principled way to do it. (Maybe expose `call` and `override` as `llm_call` and `llm_override` inside core?) 

Progress on #811

teamdandelion commented on docs/generate_provider_examples.py:
```diff
@@ -0,0 +1,115 @@
+import re
+from pathlib import Path
+
+from mirascope.llm._protocols import Provider
```
I'm wondering if there's a reason for not exposing `Provider` through a public module? Seems nice to have especially as its part of the type signature for `llm.call`

teamdandelion replied:
Decided to expose it in #826

willbakst commented on docs/generate_provider_examples.py:
```diff
@@ -0,0 +1,150 @@
+from pydantic import BaseModel
+
+EXAMPLES_ROOT = Path("examples")
+SNIPPETS_DIR = Path("build/snippets")
```
since this is ignored for git, there's a chance the folder doesn't exist.

i tried running mkdocs serve and it failed. we should create this path if it does not already exist

willbakst commented on docs/generate_provider_examples.py:
```diff
@@ -0,0 +1,150 @@
+PROVIDER_INFO: dict[Provider, ProviderInfo] = {
+    "openai": ProviderInfo(provider="openai", title="OpenAI", model="gpt-4o-mini"),
+    "anthropic": ProviderInfo(
+        provider="anthropic", title="Anthropic", model="claude-3-5-sonnet-20240620"
```
nit: we should should update this to `claude-3-5-sonnet-latest`

teamdandelion replied:
Great - updated

willbakst commented on docs/generate_provider_examples.py:
```diff
@@ -0,0 +1,150 @@
+    "mistral": ProviderInfo(
+        provider="mistral", title="Mistral", model="mistral-large-latest"
+    ),
+    "google": ProviderInfo(provider="google", title="Google", model="gemini-1.5-flash"),
```
nit: we should update this to `gemini-2.0-flash`

teamdandelion replied:
Updated

teamdandelion commented:
Pushed fixes, including tests that verify we properly create the snippets directory and bring code coverage back to 100%

teamdandelion commented on examples/learn/response_models/basic_usage/messages.py:
```diff
@@ -0,0 +1,26 @@
+print(book)
+# Output: title='The Name of the Wind' author='Patrick Rothfuss'
+
+response = cast(openai.OpenAICallResponse, book._response)  # pyright: ignore[reportAttributeAccessIssue]
```
@willbakst — do you think it's a better example to casting it as an OpenAICallResponse here, or do you think we can further simplify by casting it as `llm.CallResponse`? They both extend BaseModel so model_dump should work either way. And if we're guiding towards provider-agnosticism, probably better to guide towards llm.CallResponse in the process.

willbakst replied:
I think casting it as `llm.CallResponse` is the better cast (re: provider-agnosticism)

teamdandelion replied:
Sounds good - done.

teamdandelion commented:
Ok, I made a few more tweaks:
- Changed to provider-agnostic llm.CallResponse, but left a generalized version of the import and casting substitution logic, in case we have other examples where we do want to keep a bit of provider-specific code.
- Changed default bedrock model to `amazon.nova-lite-v1:0`, because it's a lot shorter than `anthropic.claude-3-haiku-20240307-v1:0` and will let us avoid inconsistent line-wrapping across providers

If you're okay with the defaults change, I think this should be good to merge.
